status-im/status-react
# Introduce malli library

### Summary

This PR is the foundational step to start using [malli](https://github.com/metosin/malli) in this project. It is not a small PR, and some parts of the code are not trivial. All this is necessary to guarantee our hot-reload, testing and REPL workflows are kept intact and enjoyable. Special thanks to @J-Son89, @clauxx, and @yqrashawn for helping review, test and propose improvements to this PR's branch.

Please, take in consideration we will only be able to realize malli's full power in future iterations. Do expect exciting ideas from other collaborators! Too many to spoil now ;)

For those without context: the mobile team watched a presentation about malli and went through a light RFC to put everyone on the same page, among other discussions here and there in PRs.

To keep things relatively short:

1. Unit, integration and component tests will short-circuit (fail) when inputs/outputs don't conform to their respective function schemas (CI should fail too).
2. Failed schema checks will not block the app from initializing, nor throw an exception that would trigger the LogBox. Exceptions are only thrown in the scope of automated tests.
3. There's zero performance impact in production code **because we only instrument**. Instrumentation is removed from the compiled code due to the usage of `^boolean js.goog/DEBUG`.
4. We shouldn't expect any meaningful slowdown during *development*. Perhaps in a distant and brighter future, where we have a ton more instrumentation, this could be an issue (although I've never had this problem). In any case, we shouldn't profile in debug mode, and instrumentation is disabled when debug mode is disabled.

### Demo

In this demo we see a live experience without hot-reload or REPL involved, just pressing things around. As the instrumented function (`step` component) is called, when an invalid input argument is passed, a red warning is automatically displayed, telling the developer to look at the logs. When the same function is called again with the correct arguments, the error is removed from the screen.

Without a mechanism to warn the developer, and since we don't throw exceptions to avoid the LogBox, the developer would frequently forget to fix schemas, increasing the likelihood of merging with such problems.

The automatic error removal only works if using `schema/instrument` because we can then wrap the original function using a HoF. Only in future endeavors we'll be able to achieve this effect with the macro approach, because we cannot just wrap a var and use `malli.core/=>`. We're considering submitting a PR to malli as well.

[demo-malli.webm](https://github.com/status-im/status-mobile/assets/46027/820988d8-f497-4d40-887a-5c911cc0c69f)

### What are we instrumenting in this PR?

Per our team's agreement, we're only instrumenting the bare minimum in this PR, to showcase 2 examples.

- Instrument a utility function `utils.money/format-amount` using the macro approach.
- Instrument a quo component `quo.components.counter.step.view/view` using the functional approach.

**Both approaches are useful**, the functional approach is powerful and allow us to instrument anonymous functions, like the ones we pass to subscriptions or event handlers, or the higher-order function `quo.theme/with-theme`. The macro approach is perfect for functions already defined with `defn`.

#### I evaluated the schema or function in the REPL but nothing changes

- If you evaluate the source function, you need to evaluate `schema/=>` or `schema/instrument` as well.
- Remember to *var quote* when using `schema/instrument`.
- You must call `(status-im2.setup.schema/setup!)` after any var is re-instrumented. It's advisable to add a keybinding in your editor to send this expression automatically to the CLJS REPL, or add the call at the end of the namespace you are working on (similar to how some devs add `(run-tests)` at the end of test namespaces).

### Where should schemas be defined?

For the moment, we should focus on instrumenting quo components, so define each function schema in the same namespace as the component's public `view` var.

To be specific:

- A schema used only to instrument a single function and not used elsewhere, like a quo component schema, wouldn't benefit from being defined in a separate namespace because that would force the developer to constantly open two files instead of one to check function signatures.
- A common schema reused across the repo, like `:schema.common/theme` should be registered in the global registry `schema.registry` so that consumers can just refer to it by keyword, as if it was a built-in malli schema.
- A common schema describing status-go entities like message, notification, community, etc can be stored either in the respective `src/status_im2/contexts/*` or registered globally, or even somewhere else. **This is yet to be defined**, but since I chose not to include schemas for them, we can postpone this guideline.

### How do I learn malli?

Malli's [README](https://github.com/metosin/malli) is fairly complete. The REPL can be your best friend in this journey. Ping other devs with more context, pairing can be much more effective :)

### Steps to test (devs only)

We suggest you at least checkout the branch and try to modify the `quo.components.counter.step.view/?schema` to see how malli works on a basic level. If you are a REPL user, please report if you faced any issue :raised_hands:

status: ready
